From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pravalika Gurram <pravalika.gurram@intel.com>
Date: Mon, 13 Jul 2026 21:36:44 +0530
Subject: [PATCH] drm/pagemap: folio_split callback not available for kernels
 <=6.19

Without folio_split callback, GPU SVM with THP is unsafe on older kernels:

- Reference count leaks and use-after-free bugs

Reference:
	56ef398996435
	mm/memremap: add driver callback support for folio splitting

Signed-off-by: Pravalika Gurram <pravalika.gurram@intel.com>
---
 drivers/gpu/drm/drm_pagemap.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/gpu/drm/drm_pagemap.c b/drivers/gpu/drm/drm_pagemap.c
index ee00a8a..2b295f0 100644
--- a/drivers/gpu/drm/drm_pagemap.c
+++ b/drivers/gpu/drm/drm_pagemap.c
@@ -1264,6 +1264,7 @@ static vm_fault_t drm_pagemap_migrate_to_ram(struct vm_fault *vmf)
 	return err ? VM_FAULT_SIGBUS : 0;
 }
 
+#ifndef BPM_DEV_PAGEMAP_OPS_FOLIO_SPLIT_NOT_PRESENT
 static void drm_pagemap_folio_split(struct folio *orig_folio, struct folio *new_folio)
 {
 	struct drm_pagemap_zdd *zdd;
@@ -1275,6 +1276,7 @@ static void drm_pagemap_folio_split(struct folio *orig_folio, struct folio *new_
 	zdd = folio_zone_device_data(orig_folio);
 	folio_set_zone_device_data(new_folio, drm_pagemap_zdd_get(zdd));
 }
+#endif
 
 static const struct dev_pagemap_ops drm_pagemap_pagemap_ops = {
 #ifdef BPM_DEV_PAGEMAP_OPS_FOLIO_FREE_NOT_PRESENT
@@ -1283,7 +1285,9 @@ static const struct dev_pagemap_ops drm_pagemap_pagemap_ops = {
 	.folio_free = drm_pagemap_folio_free,
 #endif
 	.migrate_to_ram = drm_pagemap_migrate_to_ram,
+#ifndef BPM_DEV_PAGEMAP_OPS_FOLIO_SPLIT_NOT_PRESENT
 	.folio_split = drm_pagemap_folio_split,
+#endif
 };
 
 /**
-- 
2.43.0

